[AArch64][GlobalISel] Fold buildvector of bitcast

davemgreen opened the pull request:
This adds a combine for buildvectors from bitcast values, sinking the bitcast and generating a buildvector from the original scalar type.
```
  %5:_(<4 x s8>) = G_BITCAST %16:_(s32)
  %18:_(s8), %19:_(s8), %20:_(s8), %21:_(s8) = G_UNMERGE_VALUES %5:_(<4 x s8>)
  %22:_(s8) = G_IMPLICIT_DEF
  %23:_(<8 x s8>) = G_BUILD_VECTOR %18:_(s8), %19:_(s8), %20:_(s8), %21:_(s8), %22:_(s8), %22:_(s8), %22:_(s8), %22:_(s8)
=>
  %undef:_(s32) = G_IMPLICIT_DEF
  %bv:_(<2 x s32>) = G_BUILD_VECTOR %16:_(s32), %undef:_(s32)
  %23:_(<8 x s8>) = G_BITCAST %bv:_(<2 x s32>)
```

It helps clean up some of the inefficiencies from widening scalar types.

arsenm commented:
I can't follow the example in the description, the register names and numbers in the result don't match the input, and there's a missing operand to the result bitcast

aemerson commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -386,6 +386,75 @@ void CombinerHelper::applyCombineConcatVectors(
 
+bool CombinerHelper::matchCombineBuildVectorOfBitcast(
+    MachineInstr &MI, SmallVector<Register> &Ops) const {
+  assert(MI.getOpcode() == TargetOpcode::G_BUILD_VECTOR &&
+         "Invalid instruction");
```
Can replace this with a `auto &BV = cast<GBuildVector>(MI);` which does an assertion for you. Then you have the `getNumSources()` and `getSourceReg(I)` helpers you can use later.

aemerson reviewed (commented):
Is there a reason we don't use `getOpcodeDef<Ty>()` here?

c-rhodes commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -386,6 +386,70 @@ void CombinerHelper::applyCombineConcatVectors(
+    if (!Op) {
+      if (!Undef)
+        Undef = Builder.buildUndef(SrcTy).getReg(0);
+      Op = Undef;
```
break?

davemgreen replied:
Multiple operands can be undef.

c-rhodes commented on llvm/test/CodeGen/AArch64/sadd_sat_vec.ll:
```diff
@@ -127,25 +127,9 @@ define void @v4i8(ptr %px, ptr %py, ptr %pz) nounwind {
-; CHECK-GI-NEXT:    mov v3.b[3], v0.b[0]
-; CHECK-GI-NEXT:    mov v5.b[3], v1.b[0]
-; CHECK-GI-NEXT:    sqadd v0.8b, v3.8b, v5.8b
+; CHECK-GI-NEXT:    ldr s0, [x0]
```
SDAG isel is quite bad here isnt it

arsenm reviewed (commented):
Reverse ping

davemgreen reviewed (commented):
Sorry for the late reply. I've updated the patch and given it a rebase.

davemgreen committed:
[AArch64][GlobalISel] Fold buildvector of bitcast

This adds a combine for buildvectors from bitcast values, sinking the bitcast
and generating a buildvector from the original scalar type.
  %5:_(<4 x s8>) = G_BITCAST %16:_(s32)
  %18:_(s8), %19:_(s8), %20:_(s8), %21:_(s8) = G_UNMERGE_VALUES %5:_(<4 x s8>)
  %22:_(s8) = G_IMPLICIT_DEF
  %23:_(<8 x s8>) = G_BUILD_VECTOR %18:_(s8), %19:_(s8), %20:_(s8), %21:_(s8), %22:_(s8), %22:_(s8), %22:_(s8), %22:_(s8)
=>
  <2 x s32> G_BUILD_VECTOR %16, %undef
  <8 x s8> G_BITCAST

It helps clean up some of the inefficiencies from widening scalar types.
